File: KotlinFlutterIntegrationTest/app/src/main/java/com/hcchiang/flutterintegrationtest/ui/flutter/FlutterIntegrateEventChannel.kt
package com.hcchiang.flutterintegrationtest.ui.flutter

import android.util.Log
import io.flutter.plugin.common.BasicMessageChannel
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.EventChannel

class FlutterIntegrateEventChannel(messenger: BinaryMessenger) : EventChannel.StreamHandler {

    private var eventChannel: EventChannel? = null

    private var eventSink: EventChannel.EventSink? = null

    private var listener: IFlutterChannelListener? = null

    init {
        try {
            eventChannel = EventChannel(messenger, EVENT_CHANNEL_NAME)
            resetHandler()
        } catch (e: Exception) {
            Log.w("AppEventChannel", "create channel has exception!!", e)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        Log.i("AppEventChannel", "onListen arguments: $arguments")
        eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        Log.i("AppEventChannel", "onCancel arguments: $arguments")
    }

    fun resetHandler() {
        cancel()
        eventChannel?.setStreamHandler(this)
    }

    fun sink(msg: String) {
        eventSink?.success(msg)
    }

    fun cancel() {
        try {
            eventSink?.endOfStream()
            eventSink = null
        } catch (e: Exception) {
            Log.w("AppEventChannel", "cancel event channel has exception!!", e)
        }
    }

    /**
     * Bind flutter message callback listener
     */
    fun setChannelListener(listener: IFlutterChannelListener) {
        this.listener = listener
    }

    fun removeChannelListener() {
        this.listener = null
    }

    companion object {
        const val EVENT_CHANNEL_NAME = "flutter_integrate_module/event"

        var instance: FlutterIntegrateEventChannel? = null
            get() = if (field != null) field else throw  ExceptionInInitializerError()
            private set

        fun setup(messenger: BinaryMessenger) {
            instance = FlutterIntegrateEventChannel(messenger)
        }
    }
}

File: SwiftFlutterIntegrateApp/SwiftFlutterIntegrateApp/FlutterControl/FlutterIntegrateEventChannel.swift
//
//  FlutterIntegrateEventChannel.swift
//  SwiftFlutterIntegrateApp
//
//  Created by Raven_Chiang on 2021/7/22.
//

import Flutter
import Foundation

class FlutterIntegrateEventChannel {
    let binaryMessenger: FlutterBinaryMessenger
    let channelName = "flutter_integrate_module/event"

    var channel: FlutterEventChannel?
    private lazy var handler: EventHandler = EventHandler()
    var countdownTimer: Timer?
    var countdown = 15

    required init(messenger: FlutterBinaryMessenger) {
        binaryMessenger = messenger
        channel = FlutterEventChannel(name: channelName, binaryMessenger: binaryMessenger)
        channel?.setStreamHandler(handler)
    }
    
    func resetHandler() {
        let newHandler = EventHandler()
        handler = newHandler
        channel?.setStreamHandler(newHandler)
    }

    func resetCountdown() {
        countdownTimer?.invalidate()
        countdownTimer = nil
        countdownTimer = Timer.scheduledTimer(withTimeInterval: 1.0, repeats: true) { timer in
            if self.countdown > 0 {
                print("\(self.countdown) seconds")
                self.countdown -= 1
                self.handler.sink(message: self.countdown)
            } else {
                timer.invalidate()
            }
        }
    }
    
    /*
     * The FlutterStreamHandler is not working with class inherit.
     * Create an instance to and pass it to [setStreamHandler] works.
     * Maybe it's because the FlutterViewController is not created in FlutterAppDelegate?
     */
    class EventHandler: NSObject, FlutterStreamHandler {
        private var channelSink: FlutterEventSink?
        
        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            print("onListen event channel")
            channelSink = events
            return nil
        }

        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            print("onCancel event channel")
            channelSink = nil
            return nil
        }
        
        func sink(message: Any?) {
            guard channelSink != nil else {
                print("FlutterEventSink is nil!!")
                return
            }
            if message != nil {
                channelSink?(message)
            }
        }
    }
}
